### PR TITLE
Initial support of getStack

### PR DESCRIPTION
**Pull request by grouma**
- Update `Location` to group JS fields into `JsLocation` and Dart fields into `DartLocation`
- Add logic to translate JS frame information into Dart frame
- Listen for pause events to cache the current Dart frame

Closes https://github.com/dart-lang/webdev/issues/348
Towards https://github.com/dart-lang/webdev/issues/352

### CODE REVIEW

File: dwds/lib/src/chrome_proxy_service.dart [hunk 1]
```diff
@@ -43,11 +44,17 @@ class ChromeProxyService implements VmServiceInterface {
...
+  ///
+  /// This is null if the [_isolate] is not paused.
+  Stack _pausedStack;
+
   /// Fields that are specific to the current [_isolate].
   ///
   /// These need to get cleared whenever [destroyIsolate] is called.
   final _classes = <String, Class>{};
   final _scriptRefs = <String, ScriptRef>{};
+  final _serverPathToScriptRef = <String, ScriptRef>{};
```
Comment (alan-knight) [inline dwds/lib/src/chrome_proxy_service.dart]: What is _scriptRefs indexed by? Can we put comments on these.  Actually, I was just looking at them and thinking we should put them in the Isolate, or some kind of Isolate subclass/wrapper. But comments would be good too.

File: dwds/lib/src/chrome_proxy_service.dart [hunk 2]
```diff
@@ -523,6 +534,9 @@ function($argsString) {
     return ScriptList()..scripts = scripts;
   }
 
+  /// Returns the corresponding [ScriptRef] for the provided Dart server path.
```
Comment (alan-knight) [inline dwds/lib/src/chrome_proxy_service.dart]: nit: for the Dart server path [uri].
and 'corresponding' isn't really necessary.

File: dwds/lib/src/chrome_proxy_service.dart [hunk 3]
```diff
@@ -771,6 +787,30 @@ function($argsString) {
     return controller;
   }
 
+  List<Frame> _translateJsFrames(DebuggerPausedEvent e) {
```
Comment (alan-knight) [inline dwds/lib/src/chrome_proxy_service.dart]: Maybe call it _dartFrames? or_dartFramesFor?

 :-) https://dart.dev/guides/language/effective-dart/design#prefer-a-noun-phrase-or-non-imperative-verb-phrase-for-a-function-or-method-if-returning-a-value-is-its-primary-purpose

File: dwds/lib/src/chrome_proxy_service.dart [hunk 4]
```diff
@@ -771,6 +787,30 @@ function($argsString) {
...
+    var dartFrames = <Frame>[];
+    var index = 0;
+    for (var frame in e.params['callFrames']) {
+      var location = frame['location'];
+      // TODO(grouma) - This function name is JS based. Add logic to
+      // translate this to a Dart function name.
+      var functionName = frame['functionName'] as String ?? '';
+      // Chrome is 0 based. Account for this.
+      var jsLocation = JsLocation(
```
Comment (alan-knight) [inline dwds/lib/src/chrome_proxy_service.dart]: To minimize the confusion, I wonder if the location objects should have constructors like .fromZeroBased that do the correction themselves. I hate having +1/-1 scattered around.

File: dwds/lib/src/debugger.dart [hunk 5]
```diff
@@ -127,9 +127,43 @@ class Debugger {
...
-      .locationsFor(uri.serverPath)
-      .firstWhere((location) => location.dartLine == line, orElse: () => null);
+  Location locationForDart(DartUri uri, int line) => sources
+      .locationsForDart(uri.serverPath)
+      .firstWhere((location) => location.dartLocation.line == line,
+          orElse: () => null);
+
+  /// Returns the closest [Location] for the given Dart source position.
+  ///
+  /// The [line] and [column] are  1-based.
```
Comment (alan-knight) [inline dwds/lib/src/debugger.dart]: extreme nit: two spaces

File: dwds/test/chrome_proxy_service_test.dart [hunk 6]
```diff
@@ -351,8 +351,30 @@ void main() {
...
+    test('returns null if not paused', () async {
+      expect(await service.getStack(isolateId), isNull);
+    });
+
+    test('returns non-empty stack when paused', () async {
+      await service.pause(isolateId);
+      // Wait for pausing to actually propagate.
+      await stream
+          .firstWhere((event) => event.kind == EventKind.kPauseInterrupted);
+      expect(await service.getStack(isolateId), isNotNull);
```
Comment (alan-knight) [inline dwds/test/chrome_proxy_service_test.dart]: Maybe a TODO for testing something about the actual values?

File: dwds/lib/src/chrome_proxy_service.dart (hunk 1 above)
Comment (grouma) [reply dwds/lib/src/chrome_proxy_service.dart]: Agree. I'll add a TODO to clean this up.

**Pull request merged**